config/config.go

```diff
@@ -133,12 +139,106 @@ func NewConfig(b []byte) (*Config, error) {
 	return conf, nil
 }
 
+// A type to wrap piv.Yubikey to manage the yubikey handle
+type YubikeyReader struct {
+	key       *piv.YubiKey
+	Overwrite bool
+}
+
+func (y *YubikeyReader) GetPIVKeyCert() (*x509.Certificate, error) {
+	if y.key == nil {
+		if err := y.connectToYubikey(); err != nil {
+			return nil, err
+		}
+	}
+	return y.key.Attest(piv.SlotSignature)
+}
+
+func (y *YubikeyReader) GenerateKey(key []byte, slot piv.Slot, opts piv.Key) (crypto.PublicKey, error) {
+	if y.key == nil {
+		if err := y.connectToYubikey(); err != nil {
+			return nil, err
+		}
+	}
+	return y.key.GenerateKey(key, slot, opts)
+}
+
+func (y *YubikeyReader) PrivateKey(slot piv.Slot, public crypto.PublicKey, auth piv.KeyAuth) (crypto.PrivateKey, error) {
+	if y.key == nil {
+		if err := y.connectToYubikey(); err != nil {
+			return nil, err
+		}
+	}
+	return y.key.PrivateKey(slot, public, auth)
+}
+
+func (y *YubikeyReader) connectToYubikey() error {
+	if y.key != nil {
+		return nil
+	}
```

Comment:
As we are just returning `nil` here, we can remove all the `if y.key == nil` checks earlier in the code.